src/ObjWriting/XModel/Export/XModelBinWriter.cpp

```diff
@@ -0,0 +1,484 @@
+                                                weights.weightCount});
+
+            vertexOffset++;
+        }
+    }
+
+    template<typename T> void Write(const T& data)
+    {
+        m_writer.write(reinterpret_cast<const char*>(&data), sizeof(T));
+    }
+
+    void WriteNullTerminatedString(std::string string)
+    {
+        m_writer.write(string.c_str(), string.size() + 1);
+    }
+
+    void WriteAlignedString(const std::string& string)
+    {
+        static constexpr uint8_t PADDING[4] = {0u, 0u, 0u, 0u};
+        const auto paddingSize = (4u - (string.size() + 1u) % 4u) % 4u;
+
+        WriteNullTerminatedString(string);
+        m_writer.write(reinterpret_cast<const char*>(PADDING), paddingSize);
+    }
+
+    void WriteComment(const std::string& comment)
+    {
+        Write(XModelBinHash::COMMENT);
+        WriteAlignedString(comment);
+    }
+
+    void WriteInt16(const int16_t hash, const int16_t value)
+    {
+        Write(hash);
+        Write(value);
+    }
+
+    void WriteUInt16(const int16_t hash, const uint16_t value)
+    {
+        Write(hash);
+        Write(value);
+    }
+
+    int16_t ClampFloatToShort(const float value) const
+    {
+        return std::clamp(static_cast<int16_t>(std::numeric_limits<int16_t>::max() * std::clamp(value, -1.0f, 1.0f)),
+                          std::numeric_limits<int16_t>::min(),
+                          std::numeric_limits<int16_t>::max());
+    }
+
+    uint8_t ClampFloatToUByte(const float value) const
+    {
+        return std::clamp(static_cast<uint8_t>(std::numeric_limits<uint8_t>::max() * std::clamp(value, -1.0f, 1.0f)),
+                          std::numeric_limits<uint8_t>::min(),
+                          std::numeric_limits<uint8_t>::max());
+    }
+
+    void WriteHeader(int16_t version)
+    {
+        WriteComment("OpenAssetTools XMODEL_BIN File");
+        WriteComment(std::format("Game Origin: {}", m_game_name));
+        WriteComment(std::format("Zone Origin: {}", m_zone_name));
+        Write(XModelBinHash::MODEL);
+        WriteInt16(static_cast<int16_t>(XModelBinHash::VERSION), version);
+    }
+
+    void WriteBones(const XModelCommon& xmodel)
+    {
+        WriteInt16(static_cast<int16_t>(XModelBinHash::BONE_COUNT), static_cast<int16_t>(xmodel.m_bones.size()));
+
+        auto boneNum = 0;
+        for (const auto& bone : xmodel.m_bones)
+        {
+            Write(XModelBinHash::BONE);
+            Write(boneNum);
+            if (bone.parentIndex)
+                Write(static_cast<int32_t>(*bone.parentIndex));
+            else
+                Write(-1);
+
+            WriteAlignedString(bone.name);
+            boneNum++;
+        }
+
+        boneNum = 0;
+        for (const auto& bone : xmodel.m_bones)
+        {
+            WriteInt16(static_cast<int16_t>(XModelBinHash::BONE_INDEX), boneNum);
```

Comment:
Why not `WriteUInt16` and making the variable be unsigned, as the boneNumber should not become negative?